query-engine/core/src/query_graph/mod.rs

```diff
@@ -831,6 +838,108 @@ impl QueryGraph {
+    /// This is only possible when the parent node _can_ fulfill the selection set.
+    /// In the case of updates and inserts, for instance, only connectors supporting `InsertReturning` and `UpdateReturning` can do it,
+    /// or else they're only able to return the primary identifier of the model inserted or updated.
+    fn normalize_data_dependencies(&mut self, capabilities: ConnectorCapabilities) -> QueryGraphResult<()> {
+        let unsatisfied_deps: Vec<(NodeRef, Vec<FieldSelection>)> = self
+            .graph
+            .node_indices()
+            .filter_map(|ix| -> Option<(NodeRef, Vec<FieldSelection>)> {
+                let node = NodeRef { node_ix: ix };
+
+                if let Node::Query(q) = self.node_content(&node).unwrap() {
+                    // If the connector does not support returning more than the primary identifier for an update,
+                    // do not update the selection set.
+                    if q.is_update_one() && !capabilities.contains(ConnectorCapability::UpdateReturning) {
+                        return None;
+                    }
+
+                    // If the connector does not support returning more than the primary identifier for a create,
+                    // do not update the selection set.
+                    if q.is_create_one() && !capabilities.contains(ConnectorCapability::InsertReturning) {
+                        return None;
+                    }
+
+                    // No connector supports returning more than the primary identifier for a delete just yet.
+                    if q.is_delete_one() {
+                        return None;
+                    }
+
+                    let edges = self.outgoing_edges(&node);
+
+                    let unsatisfied_dependencies: Vec<FieldSelection> = edges
+                        .into_iter()
+                        .filter_map(|edge| match self.edge_content(&edge).unwrap() {
+                            QueryGraphDependency::ProjectedDataDependency(ref requested_selection, _)
+                                if !q.satisfies(requested_selection) =>
+                            {
+                                trace!(
+                                    "Query {:?} does not return requested selection {:?} and will be updated.",
+                                    q,
+                                    requested_selection.prisma_names().collect::<Vec<_>>()
+                                );
+                                Some(requested_selection.clone())
+                            }
+                            _ => None,
+                        })
+                        .collect();
+
+                    if unsatisfied_dependencies.is_empty() {
+                        return None;
+                    }
+
+                    Some((node, unsatisfied_dependencies))
+                } else {
+                    None
+                }
```

Comment:
I rather some duplication than the wrong abstraction, but wonder if it would be worth to extract this calculation to a function, given that if this changes, the one in `insert_reloads` would require to change as well.

Reply:
That's a great point. `find_unsatisfied_dependencies` is a nicely extracted bit a pure logic 👍